File: java/src/main/java/io/binac/leetcode/PalindromePartitioning.java
package io.binac.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * <p>Given a string <code>s</code>, partition <code>s</code> such that every substring of the partition is a <strong>palindrome</strong>. Return all possible palindrome partitioning of <code>s</code>.</p>
 *
 * <p>A <strong>palindrome</strong> string is a string that reads the same backward as forward.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <pre><strong>Input:</strong> s = "aab"
 * <strong>Output:</strong> [["a","a","b"],["aa","b"]]
 * </pre><p><strong>Example 2:</strong></p>
 * <pre><strong>Input:</strong> s = "a"
 * <strong>Output:</strong> [["a"]]
 * </pre>
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= s.length &lt;= 16</code></li>
 * 	<li><code>s</code> contains only lowercase English letters.</li>
 * </ul>
 */
public class PalindromePartitioning {
    public static class Solution1 {
        private void partition(char[] chars, int index, boolean[][] palindrome, List<List<String>> result, List<String> path) {
            if (index == chars.length) {
                result.add(new ArrayList<>(path));
                return;
            }
            for (int j = index; j < chars.length; ++j) {
                if (palindrome[index][j]) {
                    path.add(new String(chars, index, j - index + 1));
                    partition(chars, j + 1, palindrome, result, path);
                    path.remove(path.size() - 1);
                }
            }
        }

        public List<List<String>> partition(String s) {
            final char[] chars = s.toCharArray();
            final int len = chars.length;
            boolean[][] dp = new boolean[len][len];
            for (int i = len - 1; i >= 0; --i) {
                dp[i][i] = true;
                if (i + 1 < len && chars[i] == chars[i + 1]) dp[i][i + 1] = true;
                for (int j = i + 2; j < len; ++j) {
                    if (dp[i + 1][j - 1] && chars[i] == chars[j])
                        dp[i][j] = true;
                }
            }

            List<List<String>> result = new ArrayList<>();
            partition(chars, 0, dp, result, new ArrayList<>(len));
            return result;
        }
    }
}

File: python/leetcodepy/palindrome_partitioning.py
"""
Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.

Example:

Input: "aab"
Output:
[
  ["aa","b"],
  ["a","a","b"]
]
"""
from typing import List


class Solution1:
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        size = len(s)
        dp = [[False] * size for _ in range(size)]
        for i in range(size - 1, -1, -1):
            for j in range(i, size):
                if s[i] == s[j] and (j - i < 2 or dp[i + 1][j - 1]):
                    dp[i][j] = True
        result = []
        self.__partition(s, 0, dp, result, [])
        return result

    def __partition(self, s: str, i: int, palindrome: List[List[bool]], result: List[List[str]], path: List[str]):
        if i == len(s):
            result.append(path[:])
            return
        for j in range(i, len(s)):
            if palindrome[i][j]:
                path.append(s[i:j + 1])
                self.__partition(s, j + 1, palindrome, result, path)
                path.pop()
